plugins/outputs/application_insights/application_insights.go

```diff
@@ -0,0 +1,410 @@
+package application_insights
+
+import (
+	"fmt"
+	"log"
+	"math"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/Microsoft/ApplicationInsights-Go/appinsights"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/outputs"
+)
+
+type TelemetryTransmitter interface {
+	Track(appinsights.Telemetry)
+	Close() <-chan struct{}
+}
+
+type DiagnosticsMessageSubscriber interface {
+	Subscribe(appinsights.DiagnosticsMessageHandler) appinsights.DiagnosticsMessageListener
+}
+
+type ApplicationInsights struct {
+	InstrumentationKey string
+	Timeout            internal.Duration
+	DiagMsgSubscriber  DiagnosticsMessageSubscriber
+	Transmitter        TelemetryTransmitter
+	diagMsgListener    appinsights.DiagnosticsMessageListener
+}
+
+var (
+	sampleConfig = `
+## Instrumentation key of the Application Insights resource.
+instrumentationKey = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxx"
+
+## Timeout on close. If not provided, will default to 5s. 0s means no timeout (not recommended).
+# timeout = "5s"
+`
+	is32Bit        bool
+	is32BitChecked bool
+)
+
+func (a *ApplicationInsights) SampleConfig() string {
+	return sampleConfig
+}
+
+func (a *ApplicationInsights) Description() string {
+	return "Send telegraf metrics to Azure Application Insights"
+}
+
+func (a *ApplicationInsights) Connect() error {
+	if a.Timeout.Duration == 0 {
+		a.Timeout = internal.Duration{Duration: time.Second * 5}
+	}
+
+	if a.Transmitter == nil {
+		a.useDefaultTransmitter()
+	}
+
+	if a.Transmitter != nil && a.DiagMsgSubscriber != nil {
+		a.diagMsgListener = a.DiagMsgSubscriber.Subscribe(func(msg string) error {
+			logOutputMsg("%s", msg)
+			return nil
+		})
+	}
+
+	return nil
+}
+
+func (a *ApplicationInsights) Write(metrics []telegraf.Metric) error {
+	if len(metrics) == 0 || a.Transmitter == nil || a.InstrumentationKey == "" {
+		return nil
+	}
+
+	for _, metric := range metrics {
+		if canBeTreatedAsAggregate(metric) {
+			if telemetry := createAggregateMetricTelemetry(metric); telemetry != nil {
+				a.Transmitter.Track(telemetry)
+				continue
+			}
+		}
+
+		if telemetry := createSimpleMetricTelemetry(metric); telemetry != nil {
+			a.Transmitter.Track(telemetry)
+		}
+	}
+
+	return nil
+}
+
+func (a *ApplicationInsights) Close() error {
+	if a.Transmitter == nil {
+		return nil
+	}
+
+	select {
+	case <-a.Transmitter.Close():
+		logOutputMsg("Closed")
+	case <-time.After(a.Timeout.Duration):
+		logOutputMsg("Close operation timed out after %v", a.Timeout.Duration)
+	}
+
+	if a.diagMsgListener != nil {
+		// We want to listen to diagnostic messages during closing
+		// That is why we stop listening only after Close() ends (or a timeout occurs)
+		a.diagMsgListener.Remove()
+	}
+
+	return nil
+}
+
+func init() {
+	outputs.Add("application_insights", func() telegraf.Output {
+		return &ApplicationInsights{
+			DiagMsgSubscriber: appinsightsDiagnosticsMessageSubscriber{},
```

Comment:
Is there a way to disable the diagMsgListener?
I can't find the doc about configuration, but I assume user can only specify simple data type in the configuration, thus the diagnostics listener is always created?

Reply:
It's a function callback, so you could disable it in your implementation.  The SDK provides only Add/Remove.